File: strategy/marvin/test/move/goalshot.lua
local GoalShot = Class("Test.Move.GoalShot", require "group/move/base")

local MoveToPos = require "task/movetopos"
local Pass = require "task/pass"
local World = require "../base/world"
local Robot = require "../base/robot"
local World = require "../base/world"
local G = World.Geometry
local Ball = require "observer/ball"
local Default = require "agent/attacker/default"
local Halt

local ShootGoal = require "task/shootgoal"

GoalShot.MIN_ROBOTS = 1
GoalShot.MAX_ROBOTS = 1

local TIMES = 3 -- number of goalshots per distance
local INTERVAL = 0.5

function GoalShot.canStart()
	return true
end

function GoalShot:_init()
	self._shotTime = nil
	self._distance = 0
	self._times = 0
	log("")
	log("Distance: "..tostring(G.FieldHeightHalf - self._distance))
end

function GoalShot:_canContinue()
	return true
end

function GoalShot:_update()
	if self._shotTime and (World.Ball.pos.y < -G.FieldHeightHalf or World.Ball.pos:distanceTo(World.OpponentKeeper.pos) < self._robots[1].radius + World.Ball.radius + 0.02) then
		log("Try No. "..tostring(self._times+1)..":")
		log("Ball travel time: "..tostring(World.Time - self._shotTime))
		self._shotTime = nil
		self._times = self._times + 1
		if self._times == TIMES then
			self._distance = self._distance + INTERVAL
			self._times = 0
			log("")
			log("Distance: "..tostring(G.FieldHeightHalf - self._distance))
		end
	end
end

function GoalShot:_updateTasks()
	local taskAssignments = {}
	self:_update()

	local prep = World.RefereeState == "IndirectOffensive"
	local shoot = World.RefereeState == "DirectOffensive"
	local abort = World.RefereeState == "KickoffOffensivePrepare"

	local pos = Vector(0, self._distance)
	if abort then
		self._shotTime = nil
		taskAssignments[self._robots[1]] = {class = MoveToPos, params = {pos, math.pi/2}, restart = true}
	elseif prep then
		taskAssignments[self._robots[1]] = {class = MoveToPos, params = {pos, math.pi/2}, restart = true}
	elseif Ball.isShot() then
		self._shotTime = World.Time
		taskAssignments[self._robots[1]] = {class = MoveToPos, params = {pos, math.pi/2}, restart = true}
	elseif shoot then
		taskAssignments[self._robots[1]] = {class = ShootGoal}
	else
		taskAssignments[self._robots[1]] = {class = MoveToPos, params = {pos, math.pi/2}, restart = true}
	end


	return taskAssignments, self._robots[1]
end

return GoalShot

File: strategy/typescript/glados/test/move/goalshot.ts
let GoalShot = Class("Test.Move.GoalShot", require "group/move/base")

import {MoveToPos} from "glados/task/shared/movetopos";
import * as World from "base/world";
let G = World.Geometry
import * as Ball from "glados/observer/ball";

import {ShootGoal} from "glados/task/attacker/shootgoal";

GoalShot.MIN_ROBOTS = 1
GoalShot.MAX_ROBOTS = 1

let TIMES = 3 // number of goalshots per distance
let INTERVAL = 0.5

function GoalShot.canStart () {
	return true
}

function GoalShot:_init () {
	this._shotTime = nil
	this._distance = 0
	this._times = 0
	log("")
	log("Distance: "+String(G.FieldHeightHalf - this._distance))
}

function GoalShot:_canContinue () {
	return true
}

function GoalShot:_update () {
	if (this._shotTime ? (World.Ball.pos.y < -G.FieldHeightHalf : World.Ball.pos.distanceTo(World.OpponentKeeper.pos) < this._robots[0].radius + World.Ball.radius + 0.02)) {
		log("Try No. "+String(this._times+1)+":")
		log("Ball travel time: "+String(World.Time - this._shotTime))
		this._shotTime = nil
		this._times = this._times + 1
		if (this._times == TIMES) {
			this._distance = this._distance + INTERVAL
			this._times = 0
			log("")
			log("Distance: "+String(G.FieldHeightHalf - this._distance))
		}
	}
}

function GoalShot:_updateTasks () {
	let taskAssignments = {}
	this._update()

	let prep = World.RefereeState == "IndirectOffensive"
	let shoot = World.RefereeState == "DirectOffensive"
	let abort = World.RefereeState == "KickoffOffensivePrepare"

	let pos = new Vector(0, this._distance)
	if (abort) {
		this._shotTime = nil
		taskAssignments[this._robots[0]] = {class: MoveToPos, params: {pos, Math.PI/2}, restart: true}
	} else if (prep) {
		taskAssignments[this._robots[0]] = {class: MoveToPos, params: {pos, Math.PI/2}, restart: true}
	} else if (Ball.isShot()) {
		this._shotTime = World.Time
		taskAssignments[this._robots[0]] = {class: MoveToPos, params: {pos, Math.PI/2}, restart: true}
	} else if (shoot) {
		taskAssignments[this._robots[0]] = {class: ShootGoal}
	} else {
		taskAssignments[this._robots[0]] = {class: MoveToPos, params: {pos, Math.PI/2}, restart: true}
	}


	return taskAssignments, this._robots[0]
}

return GoalShot
